lib/Feature/FeatureModelWriter.cpp

```diff
@@ -0,0 +1,347 @@
+
+#include "libxml/xmlwriter.h"
+
+#define CHECK_RC                                                               \
+  if (RC < 0) {                                                                \
+    return RC;                                                                 \
+  }
+
+namespace vara::feature {
+
+static constexpr xmlChar INDENTATIONSTRING[] = "  ";
+static constexpr char ENCODING[] = "UTF-8";
+
+int FeatureModelXmlWriter::writeFeatureModel(std::string Path) {
+  int RC;
+  std::unique_ptr<xmlTextWriter, void (*)(xmlTextWriterPtr)> Writer(
+      xmlNewTextWriterFilename(Path.data(), 0), &xmlFreeTextWriter);
+  if (!Writer) {
+    return false;
+  }
+
+  RC = writeFeatureModel(Writer.get());
+  return RC;
+}
+
+std::optional<std::string> FeatureModelXmlWriter::writeFeatureModel() {
+  int RC;
+
+  std::unique_ptr<xmlDocPtr, void (*)(xmlDocPtr *)> DocPtrPtr(
+      new xmlDocPtr(), [](xmlDocPtr *Ptr) {
+        xmlFreeDoc(*Ptr);
+        delete (Ptr);
+      });
+
+  std::unique_ptr<xmlTextWriter, void (*)(xmlTextWriterPtr)> Writer(
+      xmlNewTextWriterDoc(DocPtrPtr.get(), 0), &xmlFreeTextWriter);
+  if (Writer == nullptr) {
+    return std::nullopt;
+  }
+  RC = writeFeatureModel(Writer.get());
+  if (RC < 0) {
+    return std::nullopt;
+  }
+
+  std::unique_ptr<xmlChar *, void (*)(xmlChar **)> XmlBuffPtr(
+      new xmlChar *(), [](xmlChar **Ptr) {
+        xmlFree(*Ptr);
+        delete (Ptr);
+      });
+  int Buffersize;
+  xmlDocDumpMemoryEnc(*DocPtrPtr, XmlBuffPtr.get(), &Buffersize, ENCODING);
+  std::string Str(reinterpret_cast<char *>(*XmlBuffPtr), Buffersize);
+
+  return Str;
+}
+
+int FeatureModelXmlWriter::writeFeatureModel(xmlTextWriterPtr Writer) {
+  int RC;
+
+  RC = xmlTextWriterSetIndent(Writer, 1);
+  CHECK_RC
+  RC = xmlTextWriterSetIndentString(Writer, INDENTATIONSTRING);
+  CHECK_RC
+
+  RC = xmlTextWriterStartDocument(Writer, "1.0", ENCODING, nullptr);
+  CHECK_RC
+  RC = writeVm(Writer);
+
+  return RC;
+}
+
+int FeatureModelXmlWriter::writeVm(xmlTextWriterPtr Writer) {
+  int RC;
+
+  RC = xmlTextWriterStartElement(Writer, XmlConstants::VM);
+  CHECK_RC
+  RC = xmlTextWriterWriteAttribute(Writer, XmlConstants::NAME,
+                                   BAD_CAST Fm.getName().data());
+  CHECK_RC
+  RC = xmlTextWriterWriteAttribute(Writer, XmlConstants::ROOT,
+                                   BAD_CAST Fm.getPath().string().data());
+  CHECK_RC
+
+  RC = writeBinaryFeatures(Writer);
+  CHECK_RC
+
+  RC = writeNumericFeatures(Writer);
+  CHECK_RC
+
+  RC = writeBooleanConstraints(Writer);
+  CHECK_RC
+
+  // TODO mixed and nonNumeric constraints when supported
+
+  RC = xmlTextWriterEndDocument(Writer); // VM
+  return RC;
+}
+
+int FeatureModelXmlWriter::writeBinaryFeatures(xmlTextWriterPtr Writer) {
+  int RC;
+
+  RC = xmlTextWriterStartElement(Writer, XmlConstants::BINARYOPTIONS);
+  CHECK_RC
+
+  for (Feature *F : Fm.features()) {
+    if (F->getKind() == Feature::FeatureKind::FK_BINARY) {
```

Comment:
```suggestion
    if (isa<BinaryFeature>(F)) {
```
Don't use Kind constants directly, if they change the code breaks. User llvm:isa oder dyn_cast instead